source/base/hdf5.cc

```diff
@@ -0,0 +1,1513 @@
+    {
+      return std::vector<hsize_t>{data.size()};
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, Vector<number>>::value,
+      std::vector<hsize_t>>::type
+    get_container_dimensions(const Container<number> &data)
+    {
+      return std::vector<hsize_t>{data.size()};
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, FullMatrix<number>>::value,
+      std::vector<hsize_t>>::type
+    get_container_dimensions(const Container<number> &data)
+    {
+      return std::vector<hsize_t>{data.m(), data.n()};
+    }
+
+
+    // This function returns the pointer to the raw data of a container
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, std::vector<number>>::value,
+      unsigned int>::type
+    get_container_size(const Container<number> &data)
+    {
+      return static_cast<unsigned int>(data.size());
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, Vector<number>>::value,
+      unsigned int>::type
+    get_container_size(const Container<number> &data)
+    {
+      return static_cast<unsigned int>(data.size());
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, FullMatrix<number>>::value,
+      unsigned int>::type
+    get_container_size(const Container<number> &data)
+    {
+      return static_cast<unsigned int>(data.m() * data.n());
+    }
+
+    // This function returns the pointer to the raw data of a container
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, std::vector<number>>::value,
+      void *>::type
+    get_container_pointer(Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong
+      return data.data();
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, Vector<number>>::value,
+      void *>::type
+    get_container_pointer(Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &data[0];
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, FullMatrix<number>>::value,
+      void *>::type
+    get_container_pointer(Container<number> &data)
```

Comment:
Thinking about these `enable_if` things some more, why don't you just write this whole thing as
```
  template <typename Number>
  void *
  get_container_points (const FullMatrix<Number> &data) {...}
```
That has the exact same effect, but is far more readable :-)

Reply:
I need to use enable_if. If would be possible to get the same effect with `constexpr if` from C++17, but I guess that is not possible with deal.ii.

For example:
``` c++
if constexpr (std::is_same<Container,
                    std::vector<typename Container::value_type>>::value)
{
 return std::vector<hsize_t>{data.size()};
}
else if constexpr(std::is_same<Container,
                   FullMatrix<typename Container::value_type>>::value)
{
 return std::vector<hsize_t>{data.m(), data.n()};
}
````

Reply:
What I meant to say is this: You are defining several of these `get_container_pointer()` functions for different argument types, and that is the appropriate way to get at these pointers given that we don't have `if constexpr` at the moment (and likely not for a few years). Each of these overloads are considered by the compiler for only one specific argument type, and you make sure that this is the case via the `enable_if` trick. But you could have done this the "traditional" way as well, which I think is easier to read:
```
  template <typename Number>
  void * get_container_pointer(const FullMatrix<Number> &data) { ... };

  template <typename T>
  void * get_container_pointer(const std::vector<T> &data) { ... };

  ...maybe other overloads...
```
I think this is entirely equivalent, but much easier to read.